keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md

```diff
@@ -0,0 +1,973 @@
+similarly to how the PriorityClass Value is populated. Storing the `AllowDisruptionByPriorityGreaterThanOrEqual`
+field in the pod spec has several benefits:
+
+1. The scheduler does not need to be aware of PiorityClasses, as all relevant information is in the pod.
+2. Mutating PriorityClass objects does not impact existing pods.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+  PriorityClassName string
+  Priority *int32
+  PreemptionPolicy *PreemptionPolicy
+
++ // AllowDisruptionByPriorityGreaterThanOrEqual indicates that the PodDisruptionBudget of the pods 
++ // associated with this priority class can only be violated during scheduler preemption by other pods 
++ // with a priority value greater than or equal to AllowDisruptionByPriorityGreaterThanOrEqual. When 
++ // Priority Admission Controller is enabled, it prevents users from setting this field. 
++ // The admission controller populates this field from PriorityClassName.
++ // Defaults to Value+1 if unset.
++ // +optiona
++ AllowDisruptionByPriorityGreaterThanOrEqual *int32
+)
+```
+
+The following is an example of a `PriorityClass` where the user sets the `allowDisruptionByPriorityGreaterThanOrEqual` as 1000
+indicating that the pods corresponding to this priorty class can only be violated by the pods with a value of priority 
+greater than or equal to 1000 during the scheduler preemption process
+
+```yaml
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: low-priority
+value: 100
+globalDefault: false
+description: "This priority class should be used for XYZ service pods only."
+allowDisruptionByPriorityGreaterThanOrEqual: 1000
+```
+
+The scheduler plugin `defaultpreemption` needs to check the value set in the `AllowDisruptionByPriorityGreaterThanOrEqual` field 
+when selecting victims.
+- if the priority of the preemptor is greater than or equal to the value of `AllowDisruptionByPriorityGreaterThanOrEqual` in victim pod, 
+  the implementation will remain consistent with the existing behavior, meaning that the scheduler will try to select victims whose PDBs 
+  are not violated by preemption, but if no such victims are found, preemption will still happen and lower priority pods will be removed 
+  despite their PDBs being violated.
+- if the priority of the preemptor is less than the value of `AllowDisruptionByPriorityGreaterThanOrEqual` in victim pod, 
+  the scheduler will check if the victim' PDBs will be violated when selecting victims
+  - if violate the victims' PDBs, this victim will not be selected as candidates.
+  - if not violate the victims' PDBs, scheduler will preempt this pod via the `/evictions` endpoint. If it responds `200 OK`, 
+    it means the eviction is allowed, and the victim is deleted, similar to sending a DELETE request to the Pod URL. If it responses
+    `429 Too Many Requests`, the scheduler will output an error log and choose another victim among the candidate victims to preempt 
```

Comment:
this sounds expensive to recalculate within the scheduling cycle, as we might even have to choose victims on a different node.

Should we just abort this cycle and retry after putting the pod in backoff?

Reply:
> This sounds expensive to recalculate within the scheduling cycle, as we might even have to choose victims on a different node.

We might just have to recalculate this one node and don't need to recalculate totally again. We can reuse the `nodesToVictims`  from the `findCandidates` phase and the intermediate results from the `SelectCandidate` phase. 

> Should we just abort this cycle and retry after putting the pod in backoff?

It's OK. But assuming no other relevant events arrive, the preemptor cannot move to activeQ until it has been flushed after 30s. It's maybe to long.

Reply:
Sounds good. We might have to adjust some behavior once you have a prototype, but we can leave that for code reviews.

Reply:
I think it is a bit confusing in the KEP that we mention only code 429, when other error codes can happen as well, but I suppose we can leave that to the code reviews as well.

Reply:
Other errors should be handled like we handle any other error in scheduler: abort the scheduling cycle. No changes there.

Reply:
It might be useful to select another victim when for example pdbs are misconfigured or pods are migrating [eviction.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L215). But the error code is too generic, and I am not sure if we want to rely that much on the API for such behaviours.

In the original behaviour branch, the fallback to delete should happen at all times, unless we get 200 from eviction. To be backwards compatible and ensure best effort case.